Context/Context.php

```diff
@@ -23,31 +23,75 @@ class Context
+     * that uses this trait.
+     *
+     * This is a general purpose read-only key-value store. Read-only in the sense that adding a new value does not
+     * mutate the existing context, but returns a new Context which has the new value added.
+     *
+     * In practical terms, this is implemented as a linked list of Context instances, with each one holding a reference
+     * to the key object, the value that corresponds to the key, and an optional reference to the parent Context
+     * (i.e. the next link in the linked list chain)
+     *
+     * @var ContextKey $key The key object. Should only be null when creating an "empty" context
+     * @var mixed $value
+     * @var Context|null $parent Reference to the parent object
+     */
+    public function __construct(ContextKey $key=null, $value=null, $parent=null)
     {
         $this->key = $key;
         $this->value = $value;
         $this->parent = $parent;
     }
 
     /**
-     * @param ContextKey $key
-     * @param mixed $value
+     * This adds a k/v pair to this Context. We do this by instantiating a new Context instance with the k/v and pass
+     * a reference to $this as the "parent" creating the linked list chain.
      *
-     * @return Context
+     * The object instantiated will be of the same type as the concrete class using this trait as referenced by __CLASS__
+     *
+     * @return Context a new Context containing the k/v
+     * @suppress PhanTypeInstantiateTrait
+     * @var mixed $value
+     * @var ContextKey $key
      */
-    public static function setValue(ContextKey $key, $value, ?Context $parent=null): Context
+    public function set(ContextKey $key, $value)
     {
-        if (null === $parent) {
-            return self::$current_context = new Context($key, $value, Context::getCurrent());
-        }
+        $cls = __CLASS__;
 
-        return new Context($key, $value, $parent);
+        return new $cls($key, $value, $this);
     }
 
-    public function set(ContextKey $key, $value): Context
+    /**
+     * This is a static version of set().
+     * This is primarily useful when the caller doesn't already have a reference to a Context that they want to mutate.
+     *
+     * There are two ways to call this function.
+     * 1) With a $parent parameter:
+     *    This should be an instance of __CLASS__ which is a concrete class that uses this Trait.
+     *    Context::setValue($key, $value, $ctx) is functionally equivalent to $ctx->set($key, $value)
+     * 2) Without a $parent parameter:
+     *    In this scenario, setValue() will use the `$current_context` reference as supplied by `getCurrent()`
+     *    `getCurrent()` will always return a valid Context/__CLASS__. If one does not exist at the global scope,
+     *    an "empty" context will be created.
+     *
+     * @param ContextKey $key
+     * @param mixed $value
+     * @param Context|null $parent
+     * @suppress PhanTypeInstantiateTrait
```

Comment:
I had to add this Phan suppression in because Phan didn't like the trait call to Context in this function call.  Ideas welcome here!

Reply:
Yeah it feels like a bug in Phan to me too.

Reply:
https://github.com/phan/phan/issues/3988